apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,553 @@
+//!
+//!     shutdown_tracer_provider(); // sending remaining spans
+//! }
+//! ```
+use apollo_parser::{ast, Parser};
+use apollo_spaceport::report::{ContextualizedStats, QueryLatencyStats, StatsContext};
+use apollo_spaceport::{Reporter, ReporterGraph};
+use async_trait::async_trait;
+use once_cell::sync::OnceCell;
+use opentelemetry::{
+    global,
+    runtime::Tokio,
+    sdk,
+    sdk::export::{
+        trace::{ExportResult, SpanData, SpanExporter},
+        ExportError,
+    },
+    trace::{TraceError, TracerProvider},
+    Value,
+};
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::time::Duration;
+use tokio::task::JoinError;
+
+use crate::configuration::{SpaceportConfig, StudioGraph};
+
+pub(crate) const DEFAULT_SERVER_URL: &str = "https://127.0.0.0:50051";
+pub(crate) const DEFAULT_LISTEN: &str = "0.0.0.0:50051";
+
+/// Pipeline builder
+#[derive(Debug)]
+pub struct PipelineBuilder {
+    graph_config: Option<StudioGraph>,
+    spaceport_config: Option<SpaceportConfig>,
+    trace_config: Option<sdk::trace::Config>,
+}
+
+/// Create a new apollo telemetry exporter pipeline builder.
+pub fn new_pipeline() -> PipelineBuilder {
+    PipelineBuilder::default()
+}
+
+impl Default for PipelineBuilder {
+    /// Return the default pipeline builder.
+    fn default() -> Self {
+        Self {
+            graph_config: None,
+            spaceport_config: None,
+            trace_config: None,
+        }
+    }
+}
+
+impl PipelineBuilder {
+    /// Assign the SDK trace configuration.
+    #[allow(dead_code)]
+    pub fn with_trace_config(mut self, config: sdk::trace::Config) -> Self {
+        self.trace_config = Some(config);
+        self
+    }
+
+    /// Assign graph identification configuration
+    pub fn with_graph_config(mut self, config: &Option<StudioGraph>) -> Self {
+        self.graph_config = config.clone();
+        self
+    }
+
+    /// Assign spaceport reporting configuration
+    pub fn with_spaceport_config(mut self, config: &Option<SpaceportConfig>) -> Self {
+        self.spaceport_config = config.clone();
+        self
+    }
+
+    /// Install the apollo telemetry exporter pipeline with the recommended defaults.
+    pub fn install_batch(mut self) -> Result<sdk::trace::Tracer, ApolloError> {
+        let exporter = self.get_exporter()?;
+
+        let batch = sdk::trace::BatchSpanProcessor::builder(exporter, Tokio)
+            .with_max_queue_size(65_536)
+            .with_max_export_batch_size(65_536)
+            .build();
+
+        let mut provider_builder = sdk::trace::TracerProvider::builder().with_span_processor(batch);
+        if let Some(config) = self.trace_config.take() {
+            provider_builder = provider_builder.with_config(config);
+        }
+        let provider = provider_builder.build();
+
+        let tracer = provider.tracer("apollo-opentelemetry", Some(env!("CARGO_PKG_VERSION")));
+        // The call to set_tracer_provider() manipulate a sync RwLock.
```

Comment:
the issue is coming from something else than that lock: https://github.com/open-telemetry/opentelemetry-rust/issues/536#issuecomment-1018604343

Reply:
It's the same comment as on main and should be updated to match your analysis. I'll do that. Also, when I was digging around in the opentelemetry code I noticed that in OTLP, we eventually call this function: `build_batch_with_exporter()` which contains:`let _ = global::set_tracer_provider(provider);`

I wonder if this is prone to the same locking problem...? In fact, I noticed that there are many places in the opentelemetry implementation where set_tracer_provider() is called in this way.

Reply:
that will cause the same issue, because the hang happens when dropping the previous tracer provider (returned by `set_tracer_provider`)

Reply:
This one's difficult to resolve, since the call to set_tracer_provider() isn't within our code but is inside the opentelemetry-otlp implementation. I think we are waiting for a fix to 536.